File: iosApp/iosApp/Features/Settings/components/ChangePasswordBottomSheet.swift
//
//  ChangePasswordBottomSheet.swift
//  iosApp
//
//  Created by velkonost on 27.11.2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import SwiftUI
import SharedSDK

struct ChangePasswordBottomSheet: View {
    
    private let changePasswordState: ChangePasswordState
    
    private let onOldPasswordChanged: (String) -> Void
    private let onNewPasswordChanged: (String) -> Void
    private let onRepeatedNewPasswordChanged: (String) -> Void
    private let onChangedClick: () -> Void
    
    init(changePasswordState: ChangePasswordState, onOldPasswordChanged: @escaping (String) -> Void, onNewPasswordChanged: @escaping (String) -> Void, onRepeatedNewPasswordChanged: @escaping (String) -> Void, onChangedClick: @escaping () -> Void) {
        self.changePasswordState = changePasswordState
        self.onOldPasswordChanged = onOldPasswordChanged
        self.onNewPasswordChanged = onNewPasswordChanged
        self.onRepeatedNewPasswordChanged = onRepeatedNewPasswordChanged
        self.onChangedClick = onChangedClick
    }
    
    var body: some View {
        ZStack {
            Color.mainBackground
            VStack {
                Spacer().frame(height: 32)
                
                SingleLineTextField(
                    value: changePasswordState.oldPassword,
                    placeholderText: SharedR.strings().settings_old_password_hint.desc().localized(),
                    onValueChanged: onOldPasswordChanged
                )
                
                SingleLineTextField(
                    value: changePasswordState.newPassword,
                    placeholderText: SharedR.strings().settings_new_password_hint.desc().localized(),
                    onValueChanged: onNewPasswordChanged
                )
                
                SingleLineTextField(
                    value: changePasswordState.repeatedNewPassword,
                    placeholderText: SharedR.strings().settings_repeat_new_password_hint.desc().localized(),
                    onValueChanged: onRepeatedNewPasswordChanged
                )
                
                Spacer()
                
                AppButton(
                    labelText: SharedR.strings().settings_change_password_button.desc().localized(),
                    isLoading: false,
                    onClick: onChangedClick
                )
                
                Spacer().frame(height: 64)
                
            }
            .contentShape(Rectangle())
            .onTapGesture {
                endTextEditing()
            }
            .padding(.horizontal, 16)
        }
        .ignoresSafeArea(.all)
    }
}

File: androidApp/features/settings/src/main/kotlin/com/velkonost/getbetter/android/features/settings/components/ChangePasswordBottomSheet.kt
package com.velkonost.getbetter.android.features.settings.components

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.ExperimentalMaterialApi
import androidx.compose.material.ModalBottomSheetLayout
import androidx.compose.material.ModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.velkonost.getbetter.core.compose.components.AppButton
import com.velkonost.getbetter.core.compose.components.SingleLineTextField
import com.velkonost.getbetter.shared.features.settings.presentation.contract.ChangePasswordState
import com.velkonost.getbetter.shared.resources.SharedR
import dev.icerock.moko.resources.compose.colorResource

@OptIn(ExperimentalMaterialApi::class)
@Composable
fun ChangePasswordBottomSheet(
    modifier: Modifier = Modifier,
    changePasswordState: ChangePasswordState,
    modalSheetState: ModalBottomSheetState,
    onOldPasswordChanged: (String) -> Unit,
    onNewPasswordChanged: (String) -> Unit,
    onRepeatedNewPasswordChanged: (String) -> Unit,
    onChangedClick: () -> Unit
) {
    ModalBottomSheetLayout(
        sheetState = modalSheetState,
        sheetShape = RoundedCornerShape(topStart = 12.dp, topEnd = 12.dp),
        sheetBackgroundColor = colorResource(resource = SharedR.colors.main_background),
        sheetContent = {
            Box(
                modifier = modifier
                    .fillMaxWidth()
                    .fillMaxHeight(0.9f)
            ) {
                Column(
                    modifier = modifier
                        .fillMaxWidth()
                        .fillMaxHeight(0.9f)
                        .padding(bottom = 40.dp)
                        .padding(horizontal = 16.dp),
                    horizontalAlignment = Alignment.CenterHorizontally
                ) {
                    Spacer(modifier.height(32.dp))

                    SingleLineTextField(
                        paddingValues = PaddingValues(top = 12.dp),
                        value = changePasswordState.oldPassword,
                        placeholderText = "old pass",
                        onValueChanged = onOldPasswordChanged,
                    )

                    SingleLineTextField(
                        paddingValues = PaddingValues(top = 12.dp),
                        value = changePasswordState.newPassword,
                        placeholderText = "new pass",
                        onValueChanged = onNewPasswordChanged,
                    )

                    SingleLineTextField(
                        paddingValues = PaddingValues(top = 12.dp),
                        value = changePasswordState.repeatedNewPassword,
                        placeholderText = "repeat new pass",
                        onValueChanged = onRepeatedNewPasswordChanged,
                    )

                    Spacer(modifier.weight(1f))

                    AppButton(
//                        modifier = modifier.padding(bottom = 64.dp),
                        labelText = "change",
                        isLoading = false,
                        onClick = onChangedClick
                    )


                }
            }

        }
    ) {}
}